Delegate AssembleLinearQF to ref/serial when not impl

jeremylt opened the pull request:
This provides `AssembleLinearQF` to the GPU backends via `/cpu/self/ref/serial` so we can do diag based preconditioning for the GPU backends soon. I'll refactor `AssembleLinearDiag` downstream of this.

I also fixed an OCCA backend bug where they did not support multiple operator fields writing into the same output vector.

jeremylt commented on backends/occa/ceed-occa-restrict.okl:
```diff
@@ -61,7 +61,7 @@
     const int rng1 = toffsets[i];
     const int rngN = toffsets[i+1];
-    double value = 0.0;
+    double value = vv[i];
```
I can isolate this to a separate commit if desirable.

jeremylt commented:
I added a testing exception for OCCA. With this change, the OCCA backends attempt to run `petsc-multigrid` but fail because the OCCA backends do not currently support QFunction reuse.

This should be fixed by #305.

jeremylt commented:
A Fortran test that was passing locally is failing on Travis. Investigating, but it's probably something small.

jeremylt commented:
It was a small bug in the QFunction on the Fortran side. Fixed now.

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -304,8 +304,58 @@ int CeedOperatorAssembleLinearQFunction(CeedOperator op, CeedVector *assembled,
+    CeedChk(ierr);
+  } else { // Delegate to /cpu/self/ref/serial
+    Ceed ceedref;
+    ierr = CeedInit("/cpu/self/ref/serial", &ceedref); CeedChk(ierr);
```
It isn't the first use of this pattern (thus perhaps out of scope for this PR), but should backends be responsible for identifying their delegate (`ceedref`), instead of losing that symmetry here in interface code? It could be stored in `ceed->ceedref`.

jeremylt replied:
Backends do generally specify their own delegate and are responsible for their creation (but not destruction). A `ceed` can specify a default delegate as well as delegates for specific objects; however, delegates are treated as all-or-nothing with respect to specific objects. The code is often interdependent enough that I have been assuming that we can't mix and match functions from different `ceed`s with the same object.

```
struct Ceed_private {
  const char *resource;
  Ceed delegate;
  Ceed parent;
  objdelegate *objdelegates;
  int objdelegatecount;
...
```

We don't currently have a way for any backend to delegate only `CeedOperatorAssembleLinearQFunction` while implementing all other `CeedOperator` functions.

This is the first use of a single function delegation. Whatever we adopt here I'd like to also use for `AssembleLinearDiagonal` and `CreateElementFDMInverse`, so it seems like the right time to change it if we want to.

I can add
1) carrying around this operator+qfunction clone for the life of the operator and/or
2) letting the backend specify this 'delegate' (maybe 'fallback'? we're overusing the phrase 'delegate' to mean two/three different things here) but default to `/cpu/self/ref/serial`

if that would be more preferable/flexible/extensible/ect

jeremylt replied:
@jedbrown, I refactored to make it clear that this is different that our typical delegation mechanism and to allow backends to select something other than `/cpu/self/ref/serial` if desired. What do you think?

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -248,6 +252,50 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop, CeedOperator subop) {
+  ierr = CeedOperatorGetFallbackResource(op, &resource);
+  Ceed ceedref;
+  ierr = CeedInit(resource, &ceedref); CeedChk(ierr);
+  op->fallbackceed = ceedref;
```
Can we have one `ceed->fallback` instead of one per operator?

jeremylt replied:
Easy enough to do - it would be consistent per Ceed

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -248,6 +252,50 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop, CeedOperator subop) {
+  opref->ceed = ceedref;
+  ierr = ceedref->OperatorCreate(opref); CeedChk(ierr);
+  opref->parentceed = op->ceed;
+  op->fallback = opref;
```
Are we making hanging references above? Is it possible to infinite loop?

jeremylt replied:
I don't believe its possible to infinite loop as written. It runs Valgrind clean when I build with only the CPU backends and use a fallback rather than the `/cpu/self/ref/blocked` impl of `AssembleLinearQF`.

jedbrown replied:
What if the delegate doesn't implement the desired method and attempts to delegate to itself? This may not be possible at present, but I'd be concerned about it happening in the future.

jeremylt replied:
I put in a check that a ceed cannot delegate to itself.

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -747,6 +806,65 @@ int CeedOperatorGetData(CeedOperator op, void **data) {
+**/
+
+int CeedOperatorGetFallbackResource(CeedOperator op, const char **resource) {
+  *resource = (const char *)op->fallbackresource;
```
Documentation should be explicit that string cannot outlive the `CeedOperator`, and could become invalid if `CeedOperatorSetFallbackResource` is called.

jeremylt replied:
Updated

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -905,6 +1023,16 @@ int CeedOperatorDestroy(CeedOperator *op) {
+  if ((*op)->fallback) {
+    ierr = (*op)->qffallback->Destroy((*op)->qffallback); CeedChk(ierr);
+    ierr = CeedFree(&(*op)->qffallback); CeedChk(ierr);
+    ierr = (*op)->fallback->Destroy((*op)->fallback); CeedChk(ierr);
```
We can't we `CeedOperatorDestroy(&(*op)->fallback);`?

jeremylt replied:
The fallback contains references to memory that has been freed, so we need to only call the backend destructor code.

jeremylt commented:
@jedbrown, good to merge now? I think I addressed your comments and I think its much better code now.

I'm thinking that I should refactor FDM Element Inverse and Assemble Linear Diag in a follow-on to both this and #372 to keep the PRs smaller.

jedbrown commented on interface/ceed-operator.c:
```diff
@@ -248,6 +248,58 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop, CeedOperator subop) {
 }
 
+/**
+  @brief Duplicate a CeedOperator with a refrence Ceed to fallback for advanced
```
```suggestion
  @brief Duplicate a CeedOperator with a reference Ceed to fallback for advanced
```

jedbrown commented on include/ceed-backend.h:
```diff
@@ -65,6 +65,11 @@ CEED_EXTERN int CeedGetObjectDelegate(Ceed ceed, Ceed *delegate,
+    const char **resource);
+CEED_EXTERN int CeedSetOperatorFallbackResource(Ceed ceed,
+    const char *resource);
+CEED_EXTERN int CeedGetOperatorFallbackParentCeed(Ceed ceed, Ceed *parent);
```
Are these necessarily `Operator` fallbacks, or does `CeedGetFallbackResource` make sense?

jeremylt replied:
Currently only Operators need this fallback idea, but I think that we could more generally use this fallback idea in the future.

jeremylt replied:
Updated
